fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/bitmap/ComposedBitmapIndexAggregate.java

```diff
@@ -319,9 +321,26 @@ private static boolean separateGroupFilters(@Nonnull QueryComponent filter,
                 return Optional.empty();
             }
             // Splice the index's key between the common grouping key and the position.
-            GroupingKeyExpression groupKeyExpression = (GroupingKeyExpression)indexAggregateFunction.getOperand();
-            GroupingKeyExpression fullKey = Key.Expressions.concat(groupKeyExpression.getGroupingSubKey(), indexKey, groupKeyExpression.getGroupedSubKey())
-                    .group(groupKeyExpression.getGroupedCount());
+            // The simplest place is directly before the position.
+            // But if part of the group is a nested concat, breaking that up would need support in QueryToKeyMatcher.
+            // Moreover, the caller needs to have arranged for a compatible index to exist, which requires new support.
+            // (https://github.com/FoundationDB/fdb-record-layer/issues/1056)
+            final GroupingKeyExpression groupKey = (GroupingKeyExpression)indexAggregateFunction.getOperand();
+            int lastKeySize = 1;
+            if (groupKey.getWholeKey() instanceof ThenKeyExpression) {
+                ThenKeyExpression thenKey = (ThenKeyExpression)groupKey.getWholeKey();
+                lastKeySize = thenKey.getChildren().get(thenKey.getChildren().size() - 1).getColumnSize();
+            }
+            final ThenKeyExpression splicedKey;
+            if (lastKeySize > groupKey.getGroupedCount()) {
+                final KeyExpression wholeKey = groupKey.getWholeKey();
+                final int wholeSize = wholeKey.getColumnSize();
+                final int splicePoint = wholeSize - lastKeySize;
+                splicedKey = Key.Expressions.concat(wholeKey.getSubKey(0, splicePoint), indexKey, wholeKey.getSubKey(splicePoint, wholeSize));
```

Comment:
It's a little unclear to me why this is a legal transformation. Why is it okay if some of the grouping columns from the operand end up in the grouped columns?

Reply:
It's supposed to be an invariant that `splicePoint <= groupedCount`, so that the splice is always someplace among the grouping columns. Let me see about implementing this so that is more obviously true.

Reply:
The grouped count of `fullKey` does not change.
It should be that the splice is _always_ in the grouping columns.
Let me try to write this in a way that is more clear.